java/sage/Carny.java

```diff
@@ -1818,11 +2325,993 @@ public void windowClosing(java.awt.event.WindowEvent evt)
+      return newValue;
+    }
+
+    private WPCauseValue binaryInsertOrReturn(List<WPCauseValue> list, Airing airing, Agent agent, WPCauseValue newEntry)
+    {
+      int id = airing.getID();
+      int low = 0;
+      int high = list.size() - 1;
+
+      while (low <= high)
+      {
+        int mid = (low + high) >>> 1;
+        WPCauseValue causeValue = list.get(mid);
+        int midVal = causeValue.airing.getID();
+
+        if (midVal < id)
+          low = mid + 1;
+        else if (midVal > id)
+          high = mid - 1;
+        else
+          return causeValue; // key found
+      }
+      list.add(low, newEntry != null ? newEntry : new WPCauseValue(agent, airing)); // key not found.
+      // Return null if we added the agent since we have nothing further to do with it in that case.
+      return null;
+    }
+
+    private void binaryInsert(List<Airing> list, Airing airing)
+    {
+      int id = airing.getID();
+      int low = 0;
+      int high = list.size() - 1;
+
+      while (low <= high)
+      {
+        int mid = (low + high) >>> 1;
+        int midVal = list.get(mid).getID();
+
+        if (midVal < id)
+          low = mid + 1;
+        else if (midVal > id)
+          high = mid - 1;
+        else
+          return; // key found
+      }
+      list.add(low, airing); // key not found.
+    }
+
+    private void binaryRemove(List<Airing> list, Airing airing)
+    {
+      int id = airing.getID();
+      int low = 0;
+      int high = list.size() - 1;
+
+      while (low <= high)
+      {
+        int mid = (low + high) >>> 1;
+        int midVal = list.get(mid).getID();
+
+        if (midVal < id)
+          low = mid + 1;
+        else if (midVal > id)
+          high = mid - 1;
+        else
+        {
+          list.remove(mid);
+          return;
+        }; // key found
+      }
+      // key not found.
+    }
+  }
+
+  public class WPCauseValue
+  {
+    // We are keying on airing.
+    final private Airing airing;
+    private Agent agent;
+    private float wp;
+
+    public WPCauseValue(Agent agent, Airing airing)
+    {
+      this.airing = airing;
+      this.agent = agent;
+      this.wp = agent != null ? agent.watchProb : MIN_WP;
+    }
+
+    /**
+     * Compares the provided agent against the current agent and replaces it and the watch
+     * probability if the new agent is more desirable.
+     *
+     * @param compareAgent The agent to compare.
+     * @return <code>true</code> if the agent was replaced.
+     */
+    public boolean compareAndReplace(Agent compareAgent)
+    {
+      float compareAgentWP = compareAgent.watchProb;
+      if (agent == null || compareAgentWP > wp)
+      {
+        agent = compareAgent;
+        wp = compareAgentWP;
+        return true;
+      }
+
+      if (compareAgentWP == wp && compareAgent.isFavorite())
+      {
+        // Check for agent priority. This returns the lower priority agent.
+        if (doBattle(compareAgent, agent) == agent)
```

Comment:
Ok. That's what I was doing before, but I noticed that it didn't match what we were doing before, but maybe I didn't follow that correctly.